### PR TITLE
Add templates for configuring Docker

### PR DESCRIPTION
**Pull request by jouvin**
First attempt at providing templates to configure various parts of Docker service.

### CODE REVIEW

File: features/docker/backup.pan [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+'/software/components/filecopy/services/{/usr/sbin/backup_docker_data}' = nlist(
+                            "config", file_contents('features/docker/backup_docker_data'),
+                            "owner", "root",
+                            "perms","0755",
+                          );
+
+include {'components/cron/config'};
+
+"/software/components/cron/entries" = {
+        SELF[length(SELF)]=nlist(
```
Comment (jrha) [inline features/docker/backup.pan]: `append(SELF, ...`

File: features/docker/backup.pan [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+include {'components/filecopy/config'};
+
+'/software/components/filecopy/services/{/usr/sbin/backup_docker_data}' = nlist(
+                            "config", file_contents('features/docker/backup_docker_data'),
+                            "owner", "root",
+                            "perms","0755",
+                          );
+
+include {'components/cron/config'};
```
Comment (jrha) [inline features/docker/backup.pan]: No curlies

File: features/docker/backup.pan [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+variable DOCKER_BACKUP_MAIL_ERR=if(is_defined(DOCKER_BACKUP_MAIL_ERR)){"-mailerr "+SELF}else{""};
+
+variable DOCKER_BACKUP_COMMAND?="/usr/sbin/backup_docker_data -dir "+DOCKER_BACKUP_DIR+" "+DOCKER_BACKUP_MAIL_OK+" "+DOCKER_BACKUP_MAIL_ERR;
+
+variable DOCKER_BACKUP_FREQUENCY?="30 11 * * *";
+
+include {'components/filecopy/config'};
+
+'/software/components/filecopy/services/{/usr/sbin/backup_docker_data}' = nlist(
```
Comment (jrha) [inline features/docker/backup.pan]: `dict`

File: features/docker/backup.pan [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+variable DOCKER_BACKUP_MAIL_OK=if(is_defined(DOCKER_BACKUP_MAIL_OK)){"-mailok "+SELF}else{""};
+
+variable DOCKER_BACKUP_MAIL_ERR=if(is_defined(DOCKER_BACKUP_MAIL_ERR)){"-mailerr "+SELF}else{""};
+
+variable DOCKER_BACKUP_COMMAND?="/usr/sbin/backup_docker_data -dir "+DOCKER_BACKUP_DIR+" "+DOCKER_BACKUP_MAIL_OK+" "+DOCKER_BACKUP_MAIL_ERR;
+
+variable DOCKER_BACKUP_FREQUENCY?="30 11 * * *";
+
+include {'components/filecopy/config'};
```
Comment (jrha) [inline features/docker/backup.pan]: No curlies

File: features/docker/backup.pan [hunk 5]
```diff
@@ -0,0 +1,31 @@
+unique template features/docker/backup;
+
+variable DOCKER_BACKUP_DIR?="/backup";
+
+variable DOCKER_BACKUP_MAIL_OK=if(is_defined(DOCKER_BACKUP_MAIL_OK)){"-mailok "+SELF}else{""};
+
+variable DOCKER_BACKUP_MAIL_ERR=if(is_defined(DOCKER_BACKUP_MAIL_ERR)){"-mailerr "+SELF}else{""};
+
+variable DOCKER_BACKUP_COMMAND?="/usr/sbin/backup_docker_data -dir "+DOCKER_BACKUP_DIR+" "+DOCKER_BACKUP_MAIL_OK+" "+DOCKER_BACKUP_MAIL_ERR;
```
Comment (jrha) [inline features/docker/backup.pan]: `format(...)`

File: features/docker/core.pan [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+include {
+    if ( OS_VERSION_PARAMS['family'] == 'el' ) {
+        format('features/docker/el%s/config', OS_VERSION_PARAMS['majorversion']);
+    } else {
+        error(format('Docker configuration: unsupported OS version (%s)', OS_VERSION_PARAMS['major']));
+    };
+};
+
+# Sysconfig for the docker service
+include {'components/sysconfig/config'};
```
Comment (jrha) [inline features/docker/core.pan]: No curlies

File: features/docker/core.pan [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+'DOCKER_TMPDIR' = if(is_defined(DOCKER_DATA_DIR)){
+    DOCKER_DATA_DIR + '/tmp';
+}else{
+    null
+};
+'other_args' = DOCKER_SRV_OPTS;
+
+#Set the docker service to be running at boot
+include {'components/chkconfig/config'};
```
Comment (jrha) [inline features/docker/core.pan]: No curlies

File: features/docker/el6/config.pan [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+}
+variable DOCKER_PACKAGE ?= if ( DOCKER_OLD_PACKAGE ) {
+                               'docker-io'
+                           } else {
+                               'docker-engine'
+                           };
+
+
+# Add package
+include {'components/spma/config'};
```
Comment (jrha) [inline features/docker/el6/config.pan]: No curlies

File: features/docker/el7/config.pan [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+descr = name of the Docker package
+values = string
+default = docker-ce
+required = No
+}
+variable DOCKER_PACKAGE ?= 'docker-ce';
+
+
+# Add package
+include {'components/spma/config'};
```
Comment (jrha) [inline features/docker/el7/config.pan]: No curlies

File: features/docker/el7/config.pan [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+'/software/packages/' = pkg_repl(DOCKER_PACKAGE);
+
+
+# Sysconfig for the docker service
+include {'components/sysconfig/config'};
```
Comment (jrha) [inline features/docker/el7/config.pan]: No curlies

File: features/docker/el7/config.pan [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+'DOCKER_TMPDIR' = if(is_defined(DOCKER_DATA_DIR)){
+    DOCKER_DATA_DIR + '/tmp';
+}else{
+    null
+};
+'other_args' = DOCKER_SRV_OPTS;
+
+#Set the docker service to be running at boot
+include {'components/chkconfig/config'};
```
Comment (jrha) [inline features/docker/el7/config.pan]: No curlies

File: features/docker/pipework.pan [hunk 12]
```diff
@@ -0,0 +1,11 @@
+unique template features/docker/pipework;
+
+#Copy some needed files
+include {'components/filecopy/config'};
```
Comment (jrha) [inline features/docker/pipework.pan]: No curlies

File: features/docker/pipework.pan [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+
+prefix '/software/components/filecopy/services/';
+
+'{/usr/bin/pipework}/config' = file_contents('features/docker/pipework');
```
Comment (jrha) [inline features/docker/pipework.pan]: Why not put the path up with the `prefix`?

File: features/docker/config.pan [hunk 14]
```diff
@@ -12,15 +12,15 @@ required = No
...
-include if( DOCKER_PIPEWORK ) 'features/docker/pipework';
+include if (DOCKER_PIPEWORK) 'features/docker/pipework';
 
 # Configure Docker YUM repository if necessary
 variable SITE_REPOSITORY_LIST = {
-    if ( is_defined(DOCKER_YUM_REPOSITORY) ) {
-        SELF[length(SELF)] = format('%s', DOCKER_YUM_REPOSITORY);
+    repolist = SELF;
+    if (is_defined(DOCKER_YUM_REPOSITORY)) {
+        repolist = append(repolist, to_string(DOCKER_YUM_REPOSITORY));
```
Comment (stdweird) [inline features/docker/config.pan]: or just `append`, no temp variables or anything (but i know you don't like that way of using append)

File: features/docker/pipework [hunk 15]
```diff
@@ -0,0 +1,248 @@
+#!/bin/sh
```
Comment (stdweird) [inline features/docker/pipework]: did you write this? no license, no author (and no rpm?)

**Comment by jrha** [comment]
@jouvin I've made changes to your branch to address my own comments, panlint and shellcheck issues.
@stdweird's question of licensing of the scripts is important and will block this being merged as we cannot redistribute software of unknown license or origin.

File: features/docker/pipework (hunk 15 above)
Comment (jrha) [reply features/docker/pipework]: Looks a lot like an earlier version of https://github.com/jpetazzo/pipework/blob/master/pipework to me.

Comment (stdweird) [reply features/docker/pipework]: that's licensed under APL2.0, so should be fine to add the source and mention it in the header that is apl

**Comment by jouvin** [comment]
I'm a bit lost about what I need to do, if anything... Is the licensing issue the only thing remaining?

**Comment by jouvin** [comment]
I really discovered this issue with the PR, I completely overlooked it! I don't know anything about the script causing the problem. May be @sartiran has an idea?

**Comment by sartiran** [comment]
Hi, I've to say I've no idea of which script you are talking about. Can you give me some more details?

**Comment by jouvin** [comment]
Hopefully, my last commit clarifies the licensing.

**Comment by jrha** [comment]
As discussed at today's standup, this still has some outstanding lint issues, if they are quickly fixed this can be merged for 18.6.

**Comment by jouvin** [comment]
Yes, I completely missed how bad the state of the templates was... I just fixed it, took quite some time. They are used at GRIF&LAL, so I'm sure the contents is correct (we had a few fixes added since the PR was open). I'm waiting for the tests to see if panlint complains (I guess it will...) and I'll try to fix the remaining details later today.

**Comment by jouvin** [comment]
panlint is happy 😄 and me too 😄

**Pull request merged**